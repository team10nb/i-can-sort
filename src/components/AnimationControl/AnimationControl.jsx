import * as React from "react";
import Zoom from "@material-ui/core/Zoom";
import SpeedMenu from "../SpeedMenu/SpeedMenu";
import Tooltip from "@material-ui/core/Tooltip";
import ReplayIcon from "@material-ui/icons/Replay";
import { makeStyles } from "@material-ui/core/styles";
import IconButton from "@material-ui/core/IconButton";
import SkipNextIcon from "@material-ui/icons/SkipNext";
import SkipPreviousIcon from "@material-ui/icons/SkipPrevious";
import PlayCircleFilledIcon from "@material-ui/icons/PlayCircleFilled";
import PauseCircleFilledIcon from "@material-ui/icons/PauseCircleFilled";

export default function AnimationControl(props) {

    const useStyles = makeStyles({
        customTooltip: {
            fontFamily: "inherit",
            fontSize: "0.5em",
            fontWeight: "700",
            paddingTop: "8px",
            paddingBottom: "10px",
            letterSpacing: "1px",
            marginTop: "3px"
        },
    });

    const classes = useStyles();
    const { handleResetClick, stepForward, stepBackward, pause, resume, isPlaying, playDisabled, backwardDisabled, handleClick, handleClose, anchorEl, playSpeed} = props;


    return (
        <div
            style={{
                position: "relative",
                left: "50%",
                transform: "translate(-50%)",
                width: "240px",
            }}
        >
            {/* reset button */}
            <Tooltip
                title='Reset'
                TransitionComponent={Zoom}
                enterDelay={300}
                leaveDelay={200}
                className={classes.customTooltip }
                arrow
            >
                <span>
                    <IconButton
                        // pause the animation and reset
                        onClick={handleResetClick}
                        disabled={false}
                    >
                        <ReplayIcon
                            // color may need to follow the theme color
                            style={{ color: "grey" }}
                            fontSize='small'
                        />
                    </IconButton>
                </span>
            </Tooltip>

            {/* previous step button */}
            <IconButton
                onClick={() => {
                    stepBackward();
                }}
                disabled={backwardDisabled}
            >
                <SkipPreviousIcon
                    // color may need to follow the theme color
                    style={{ color: "grey" }}
                    fontSize='small'
                />
            </IconButton>

            {/* play button */}
            <IconButton
                // function using depends on isPlaying
                onClick={() => {
                    isPlaying ? pause() : resume();
                }}
                disabled={playDisabled}
            >
                {/* button appearence depends on isPlaying */}
                {isPlaying ? (
                    <PauseCircleFilledIcon fontSize='large' />
                ) : (
                    <PlayCircleFilledIcon fontSize='large' />
                )}
            </IconButton>

            {/* next step button */}
            <IconButton
                onClick={() => {
                    stepForward();
                }}
                disabled={playDisabled}
            >
                <SkipNextIcon
                    // color may need to follow the theme color
                    style={{ color: "grey" }}
                    fontSize='small'
                />
            </IconButton>

            {/* the speed choosing menu */}
            <SpeedMenu
                handleClick={handleClick}
                handleClose={handleClose}
                anchorEl={anchorEl}
                speed={playSpeed + "x"}
            />
        </div>
    );
}
