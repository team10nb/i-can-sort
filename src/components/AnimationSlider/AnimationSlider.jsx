import React from "react";
import { makeStyles, withStyles } from "@material-ui/core/styles";
import Slider from "@material-ui/core/Slider";

const PrettoSlider = withStyles({
    root: {
        color: "#52af77",
        height: 4,
        alignContent: "center",
    },
    thumb: {
        height: 24,
        width: 24,
        backgroundColor: "#fff",
        border: "2px solid currentColor",
        marginTop: -8,
        marginLeft: -12,
        "&:focus, &:hover, &$active": {
            boxShadow: "inherit",
        },
    },
    active: {},
    valueLabel: {
        left: "calc(-50% + 4px)",
    },
    track: {
        height: 5,
        borderRadius: 2,
    },
    rail: {
        height: 5,
        borderRadius: 2,
    },
    mark: {
        height: 5,
    }
})(Slider);

export default function DiscreteSlider(props) {
    const { width, step, max, handleChange, value } = props;

    const useStyles = makeStyles({
        root: {
            width: "500px",
            marginTop: "10px",
        },
    });
    const classes = useStyles();

    return (
        <div className={classes.root}>
            <PrettoSlider
                value={typeof value === "number" ? value : 0}
                aria-labelledby='Progress'
                valueLabelDisplay='auto'
                step={step}
                marks
                min={0}
                max={max}
                onChange={handleChange}
            />
        </div>
    );
}
