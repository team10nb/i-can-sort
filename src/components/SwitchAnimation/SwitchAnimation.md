### **Developer**

Ruizi Han, Shiliang Chen

Reference: MATERIAL-UI (https://material-ui.com)

### **Introduction**

Animation component which contains a control bar, a slider and a description box.

### **Import**

```html
import Zoom from "@material-ui/core/Zoom"; 
import SpeedMenu from "../SpeedMenu/SpeedMenu"; 
import Tooltip from "@material-ui/core/Tooltip";
import ReplayIcon from "@material-ui/icons/Replay"; 
import { makeStyles } from "@material-ui/core/styles"; 
import IconButton from "@material-ui/core/IconButton";
import SkipNextIcon from "@material-ui/icons/SkipNext"; 
import SkipPreviousIcon from "@material-ui/icons/SkipPrevious"; 
import PlayCircleFilledIcon from "@material-ui/icons/PlayCircleFilled"; 
import PauseCircleFilledIcon from "@material-ui/icons/PauseCircleFilled";
```

### **Props**

| Name             | Type   | Default | Description                                                   |
| ---------------- | ------ | ------- | ------------------------------------------------------------- |
| width        | number | 0       | The width of this component                   |
| trace         | object | null    | Indicate the sorting process                             |
| description         | [string] | []    | Contain array of descriptions                             |

### **Demo**

```jsx
import BubbleSort from "../SwitchAnimation/SortingAlgorithms/BubbleSort/BubbleSort";

const trace = BubbleSort([2, 1, 6, 10, 9, 12, 8, 3]);
<SwitchAnimation {...trace} width={600} />
```
