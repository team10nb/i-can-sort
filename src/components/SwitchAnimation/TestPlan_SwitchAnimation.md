## Test Plan
--------
<br>


#### Component: SwitchAnimation

|Test Name|Description|Operation|Expected Outcome|
|----|------|----------------|------------|
|should call AnimationSlider|Test whether AnimationSlider is successfully called|Mock AnimationSlider and call it|call time is 1|
|should call AnimationControl|Test whether AnimationControl is successfully called|Mock AnimationControl and call it|call time is 1|
| should match Snapshot | Compare new snapshot with old one. | Take a snapshot and compare | Snapshots are the same. |


