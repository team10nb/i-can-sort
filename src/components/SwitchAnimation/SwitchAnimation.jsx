/*
    Author: Shiliang Chen, Ruizi Han

    The animation of bars
*/

import "@fontsource/roboto";
import * as React from "react";
import { motion } from "framer-motion";
import { useState, useEffect } from "react";
import AnimationSlider from "../AnimationSlider/AnimationSlider";
import { makeStyles } from "@material-ui/core/styles";
import AnimationControl from "../AnimationControl/AnimationControl";
import ExplanationBox from "../ExplanationBox/ExplanationBox";
import ArrowDropUpIcon from '@material-ui/icons/ArrowDropUp';

// a framer motion transition attributes
const spring = {
    type: "spring", // a framer motion type that simulates spring
    damping: 15, //Strength of opposing force. If set to 0, spring will oscillate indefinitely
    stiffness: 150, //Stiffness of the spring. Higher values will create more sudden movement. Set to 100 by default.
    mass: 0.1, // Mass of the moving object. Higher values will result in more lethargic movement
};

export const SwitchAnimation = (props) => {
    const { trace, description, width, explanationBoxHeight } = props;

    // The bars displayed to visulise the numbers
    const [bars, setBars] = useState(trace[0]);
    // The speed of playing the animation
    const [playSpeed, setPlaySpeed] = useState(1);
    // The current step among traces
    const [currentStep, setCurrentStep] = useState(0);
    // It is used to clean timeouts to pause
    const [timeOutIds, setTimeOutIds] = useState([]);
    // The state of the animation, whether is playing
    const [isPlaying, setIsPlaying] = useState(false);
    // It is used to control the speed menu
    const [anchorEl, setAnchorEl] = useState(null);
    // The state of play button and next step button, whether is disabled
    const [playDisabled, setPlayDisabled] = useState(false);
    // the State of previous step bnutton
    const [backwardDisabled, setBackwardDisabled] = useState(true);

    const useStyles = makeStyles({
        root: {
            display: "grid",
            justifyContent: "center",
            alignContent: "flex-end",
            height: 430,
            marginTop: "20px",
        },
        bars: {
            listStyle: "none",
            padding: 0,
            margin: 0,
            marginBottom: 15,
            position: "relative",
            display: "flex",
            flexWrap: "wrap-reverse",
            justifyContent: "center",
        },
        bar: {
            listStyle: "none",
            padding: 0,
            margin: 0,

            boxShadow: "0px 0px 2px 2px #88888833",
            borderRadius: "10px",
            marginBottom: "10px",
            marginRight: "15px",
            marginLeft: "15px",
            width: "30px",
            height: "140px",
            fontSize: "0.1em",
            justifyContent: "center",
        },
        barNumber: {
            fontSize: "16px",
            marginTop: "-23px",
            textAlign: "center",
            fontWeight: "600",
        },
    });

    const classes = useStyles();

    // Update buttons' disable property when steps are changed
    useEffect(() => {
        currentStep === 0
            ? setBackwardDisabled(true)
            : setBackwardDisabled(false);
        currentStep + 1 === trace.length
            ? setPlayDisabled(true)
            : setPlayDisabled(false);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [currentStep]);

    // Use the latest speed to play the animation
    useEffect(() => {
        // if it is playing, replay
        if (isPlaying) {
            pause();
            resume();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [playSpeed]);

    useEffect(() => {
        handleResetClick();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [trace]);

    // It is used to open the speed menu
    const handleClick = (event) => {
        setAnchorEl(event.currentTarget);
    };

    // It is used to adjust speed, triggered when close the speed menu
    const handleClose = (event) => {
        const value = event.nativeEvent.target.value / 4;
        if (!isNaN(value)) {
            setPlaySpeed(value);
        }
        setAnchorEl(null);
    };

    const handleSliderChange = (event, newValue) => {
        if (isPlaying) {
            pause();
        }

        const item = trace[newValue];
        setCurrentStep(newValue);
        setBars(item);
    };

    // It is used to clean timeouts to pause the animation
    const clearTimeouts = () => {
        timeOutIds.forEach((timeoutId) => clearTimeout(timeoutId));
        setTimeOutIds([]);
    };

    // Main method, used to generate a time line based on the speed
    // change bars after a time interval
    const run = (trace) => {
        // The current bars are the [0] of trace, to make the animation start without delay
        // here we begin the animation from [1] of trace
        // it's also suitable for pause, step forward and backwar
        const subTrace = trace.slice(1);
        const timeoutIds = [];
        // a time interval unit
        const timer = 1200 / playSpeed;

        // Set a timeout for each item in the trace
        subTrace.forEach((item, i) => {
            let timeoutId = setTimeout(
                (item) => {
                    // update the current step
                    setCurrentStep((prevStep) =>
                        i === trace.length - 1 ? prevStep : prevStep + 1
                    );
                    // update bars to be animated
                    setBars(item);
                    i === subTrace.length - 1
                        ? setIsPlaying(false)
                        : setIsPlaying(true);
                },
                i * timer, //time interval
                item
            );
            timeoutIds.push(timeoutId);
        });

        // Clear timeouts upon completion
        let timeoutId = setTimeout(clearTimeouts, trace.length * timer);
        timeoutIds.push(timeoutId);
        setTimeOutIds(timeoutIds);
    };

    // To pause the animation
    const pause = () => {
        setIsPlaying(false);
        clearTimeouts();
    };

    // To resume the animation
    const resume = () => {
        setIsPlaying(true);
        const newtrace = trace.slice(currentStep);
        run(newtrace);
    };

    // Go to next step and pause
    const stepForward = () => {
        if (currentStep < trace.length - 1) {
            pause();
            const item = trace[currentStep + 1];
            setCurrentStep((prevStep) => prevStep + 1);
            setBars(item);
        }
    };

    // Go to the previous step and pause
    const stepBackward = () => {
        if (currentStep > 0) {
            pause();
            const item = trace[currentStep - 1];
            setCurrentStep((prevStep) => prevStep - 1);
            setBars(item);
        }
    };

    const handleResetClick = () => {
        pause();
        setCurrentStep(0);
        setBars(trace[0]);
    };

    const animationControlProps = {
        handleResetClick,
        stepForward,
        stepBackward,
        pause,
        resume,
        isPlaying,
        playDisabled,
        backwardDisabled,
        handleClick,
        handleClose,
        anchorEl,
        playSpeed,
        trace,
    };

    return (
        <div className={classes.root}>
            {/* bars */}
            <div className={classes.bars}>
                {bars.map((bar) => (
                    <motion.li
                        key={bar.key} // each bar's identification
                        layout
                        transition={spring}
                        style={bar}
                        className={classes.bar}
                        animate={{
                            backgroundColor: bar.backgroundColor,
                            y: bar.y,
                            x: bar.x,
                        }}
                    >
                        <div className={classes.barNumber}>
                            {bar.value}
                        </div>
                        <div
                            style={{
                                marginTop: bar.height - 17,
                                fontSize: "16px",
                                textAlign: "center",
                                fontWeight: "600",
                            }}
                        >
                            {bar.isPivot ? <ArrowDropUpIcon style={{ marginBottom: "-18px" }} /> : ""}
                        </div>
                    </motion.li>
                ))}
            </div>
            <ExplanationBox width={60} height={explanationBoxHeight}>
                {description[currentStep]}
            </ExplanationBox>

            <AnimationSlider
                width={width}
                step={1}
                max={trace.length - 1}
                handleChange={handleSliderChange}
                value={currentStep}
                display="none"
            />

            <AnimationControl {...animationControlProps} />
        </div>
    );
};

export default SwitchAnimation;
