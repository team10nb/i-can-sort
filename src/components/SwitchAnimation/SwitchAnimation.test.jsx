import { render } from "@testing-library/react";
import SwitchAnimation from "./SwitchAnimation";
import * as AnimationSliders from "../AnimationSlider/AnimationSlider";
import * as AnimationControls from "../AnimationControl/AnimationControl";

test('should call AnimationSlider', () => {
    const sliderSpy = jest.spyOn(AnimationSliders, "default").mockImplementation(() => <div>mockAnimationSlider</div>);

    const {getByText} = render(<SwitchAnimation />);
    getByText("mockAnimationSlider");
    expect(sliderSpy).toBeCalledTimes(1);
})

test('should call AnimationControl', () => {
    jest.restoreAllMocks();
    const controlSpy = jest.spyOn(AnimationControls, "default").mockImplementation(() => <div>mockAnimationControl</div>);

    const {getByText} = render(<SwitchAnimation />);
    getByText("mockAnimationControl");
    expect(controlSpy).toBeCalledTimes(1);
});

test('should match snapshot', () => {
    jest.restoreAllMocks();
    // jest.resetModules();
    const content = render(<SwitchAnimation />);

    expect(content).toMatchSnapshot();
})
