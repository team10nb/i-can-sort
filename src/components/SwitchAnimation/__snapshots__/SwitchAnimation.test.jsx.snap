// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-root-72"
      >
        <div
          class="makeStyles-bars-73"
        >
          <li
            class="makeStyles-bar-74"
            style="background-color: rgb(255, 0, 140); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
          >
            <div
              class="makeStyles-barNumber-75"
            />
            <div
              style="margin-top: 33px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
          <li
            class="makeStyles-bar-74"
            style="background-color: rgb(211, 9, 225); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
          >
            <div
              class="makeStyles-barNumber-75"
            />
            <div
              style="margin-top: 13px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
          <li
            class="makeStyles-bar-74"
            style="background-color: rgb(156, 26, 255); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
          >
            <div
              class="makeStyles-barNumber-75"
            />
            <div
              style="margin-top: 73px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
          <li
            class="makeStyles-bar-74"
            style="background-color: rgb(119, 0, 255); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
          >
            <div
              class="makeStyles-barNumber-75"
            />
            <div
              style="margin-top: -7px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
          <li
            class="makeStyles-bar-74"
            style="background-color: rgb(102, 170, 0); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
          >
            <div
              class="makeStyles-barNumber-75"
            />
            <div
              style="margin-top: 3px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
        </div>
        <div
          class="makeStyles-root-76"
        >
          <div
            class="MuiPaper-root WithStyles(ForwardRef(Paper))-root-53 MuiPaper-outlined MuiPaper-rounded"
          />
        </div>
        <div
          class="makeStyles-root-77"
        >
          <span
            class="MuiSlider-root WithStyles(ForwardRef(Slider))-root-55 MuiSlider-colorPrimary"
          >
            <span
              class="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-60"
            />
            <span
              class="MuiSlider-track WithStyles(ForwardRef(Slider))-track-59"
              style="left: 0%; width: 0%;"
            />
            <input
              type="hidden"
              value="0"
            />
            <span
              class="MuiSlider-mark WithStyles(ForwardRef(Slider))-mark-61 MuiSlider-markActive"
              data-index="0"
              style="left: 0%;"
            />
            <span
              class="MuiSlider-mark WithStyles(ForwardRef(Slider))-mark-61"
              data-index="1"
              style="left: 100%;"
            />
            <span
              aria-labelledby="Progress"
              aria-orientation="horizontal"
              aria-valuemax="1"
              aria-valuemin="0"
              aria-valuenow="0"
              class="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-56 MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-62"
              data-index="0"
              role="slider"
              style="left: 0%;"
              tabindex="0"
            >
              <span
                class="PrivateValueLabel-offset-64 MuiSlider-valueLabel WithStyles(ForwardRef(Slider))-valueLabel-58"
              >
                <span
                  class="PrivateValueLabel-circle-65"
                >
                  <span
                    class="PrivateValueLabel-label-66"
                  >
                    0
                  </span>
                </span>
              </span>
            </span>
          </span>
        </div>
        <div
          style="position: relative; left: 50%; transform: translate(-50%); width: 240px;"
        >
          <span
            class="makeStyles-customTooltip-78"
            title="Reset"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  style="color: grey;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </span>
          <button
            class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                style="color: grey;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
                />
              </svg>
            </span>
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 14.5v-9l6 4.5-6 4.5z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                style="color: grey;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <div
            class="makeStyles-root-79"
          >
            <button
              aria-controls="simple-menu"
              aria-haspopup="true"
              class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-81"
              style="min-width: 40px;"
              tabindex="0"
              title="Speed"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                1x
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="MuiPopover-root makeStyles-menu-80"
      id="simple-menu"
      role="presentation"
      style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px; visibility: hidden;"
    >
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
        style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
        tabindex="-1"
      >
        <ul
          class="MuiList-root MuiMenu-list MuiList-padding"
          role="menu"
          tabindex="-1"
        >
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-83 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="0"
            value="1"
          >
            0.25x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-83 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="2"
          >
            0.5x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-83 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="4"
          >
            1x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-83 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="8"
          >
            2x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-83 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="16"
          >
            4x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-root-72"
    >
      <div
        class="makeStyles-bars-73"
      >
        <li
          class="makeStyles-bar-74"
          style="background-color: rgb(255, 0, 140); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
        >
          <div
            class="makeStyles-barNumber-75"
          />
          <div
            style="margin-top: 33px; font-size: 16px; text-align: center; font-weight: 600;"
          />
        </li>
        <li
          class="makeStyles-bar-74"
          style="background-color: rgb(211, 9, 225); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
        >
          <div
            class="makeStyles-barNumber-75"
          />
          <div
            style="margin-top: 13px; font-size: 16px; text-align: center; font-weight: 600;"
          />
        </li>
        <li
          class="makeStyles-bar-74"
          style="background-color: rgb(156, 26, 255); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
        >
          <div
            class="makeStyles-barNumber-75"
          />
          <div
            style="margin-top: 73px; font-size: 16px; text-align: center; font-weight: 600;"
          />
        </li>
        <li
          class="makeStyles-bar-74"
          style="background-color: rgb(119, 0, 255); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
        >
          <div
            class="makeStyles-barNumber-75"
          />
          <div
            style="margin-top: -7px; font-size: 16px; text-align: center; font-weight: 600;"
          />
        </li>
        <li
          class="makeStyles-bar-74"
          style="background-color: rgb(102, 170, 0); transform: translate3d(0px, 0px, 0) scale(NaN, NaN); transform-origin: 50% 50% 0;"
        >
          <div
            class="makeStyles-barNumber-75"
          />
          <div
            style="margin-top: 3px; font-size: 16px; text-align: center; font-weight: 600;"
          />
        </li>
      </div>
      <div
        class="makeStyles-root-76"
      >
        <div
          class="MuiPaper-root WithStyles(ForwardRef(Paper))-root-53 MuiPaper-outlined MuiPaper-rounded"
        />
      </div>
      <div
        class="makeStyles-root-77"
      >
        <span
          class="MuiSlider-root WithStyles(ForwardRef(Slider))-root-55 MuiSlider-colorPrimary"
        >
          <span
            class="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-60"
          />
          <span
            class="MuiSlider-track WithStyles(ForwardRef(Slider))-track-59"
            style="left: 0%; width: 0%;"
          />
          <input
            type="hidden"
            value="0"
          />
          <span
            class="MuiSlider-mark WithStyles(ForwardRef(Slider))-mark-61 MuiSlider-markActive"
            data-index="0"
            style="left: 0%;"
          />
          <span
            class="MuiSlider-mark WithStyles(ForwardRef(Slider))-mark-61"
            data-index="1"
            style="left: 100%;"
          />
          <span
            aria-labelledby="Progress"
            aria-orientation="horizontal"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="0"
            class="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-56 MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-62"
            data-index="0"
            role="slider"
            style="left: 0%;"
            tabindex="0"
          >
            <span
              class="PrivateValueLabel-offset-64 MuiSlider-valueLabel WithStyles(ForwardRef(Slider))-valueLabel-58"
            >
              <span
                class="PrivateValueLabel-circle-65"
              >
                <span
                  class="PrivateValueLabel-label-66"
                >
                  0
                </span>
              </span>
            </span>
          </span>
        </span>
      </div>
      <div
        style="position: relative; left: 50%; transform: translate(-50%); width: 240px;"
      >
        <span
          class="makeStyles-customTooltip-78"
          title="Reset"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                style="color: grey;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </span>
        <button
          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              style="color: grey;"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
              />
            </svg>
          </span>
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 14.5v-9l6 4.5-6 4.5z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              style="color: grey;"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="makeStyles-root-79"
        >
          <button
            aria-controls="simple-menu"
            aria-haspopup="true"
            class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-81"
            style="min-width: 40px;"
            tabindex="0"
            title="Speed"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              1x
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
